### PR TITLE
[Rest Server] Add pod priority to support FIFO job scheduling

### LINKED ISSUE #3704: P1 - Job Near FIFO scheduling
**What would you like to be added**:

**Why is this needed**:

**Without this feature, how does the current module work**:

**Components that may involve changes**:

**Potential Solutions**:
1. [**LongTerm**][Strict FIFO]: Support workflow/jobdependeny by user or PAI (such as Aether https://github.com/microsoft/pai/issues/3702) - **Pending**
2. [**Picked**][Near FIFO]: RestServer leverage [Pod priority](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/) to nearly ensure Job FIFO scheduling in default scheduler - **Works: near FIFO, need to disable Priority Admission Controller and set per job priority decreasingly, [disable k8s preemption](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#how-to-disable-preemption), a little hack and [small chances to starve later jobs](https://github.com/kubernetes/kubernetes/issues/71486#issuecomment-445567835)**
Such as PodPriority=-(MaxJobPriority-JobPriority, JobCreationTime)
     ```
     Given JobPriority range: [-1, 126]
     PodPriority int32bits =
     FixToNegative: uint1bits = (Fix to 1) | 
     JobPriority: uint8bits = (126 - JobPriority) | 
     JobCreationTime: uint23bits = (UnixEpoch % (2^23))
     ```
   This makes sure job Pod priority is by default lower than system services Pod priority, and k8s reserved priorities.
3. [**LongTerm**][Near FIFO]: Changes for default scheduler to do FIFO queuing according to the Pod creation time instead of the Pod last queued time by scheduler. - **Long term to be on AKS and [small chances to starve later jobs](https://github.com/kubernetes/kubernetes/issues/71486#issuecomment-445567835) and cannot tolerate job interleave such as task retry**
Tracked in https://github.com/kubernetes/kubernetes/issues/83834 and fixed by https://github.com/kubernetes/kubernetes/pull/83835.
4. [Infeasible][Weak FIFO]: Leverage existing default scheduler FIFO queue - **Not work: Very weak FIFO**
5. [Infeasible][Weak FIFO]: Leverage existing default scheduler Priority queue - **Not work: Very weak FIFO**

**Tests**:
Setup scheduler with No HA and No Restart.
- [x] Priority queue + Decreasing Pod Priority: **Works: near FIFO** (<5% cut in line)
- [x] Priority queue + Same Pod Priority: **Not work** (>40% cut in line)
- [x] Priority queue + Same Pod Priority + Using Pod creation time to sort: **Works: near FIFO** (<5% cut in line)
- [x] FIFO queue: **Not work** (>40% cut in line)
image
image
image
image

### PR DESCRIPTION
**Pull request by abuccts**
Add pod priority to support FIFO job scheduling.

Closes #3704.

### CODE REVIEW

File: src/rest-server/src/middlewares/v2/hived.js [hunk 1]
```diff
@@ -24,8 +24,8 @@ const vcConfig = require('@pai/config/vc');
 const convertPriority = (priorityClass) => {
   // TODO: make it a cluster-wise config
   const priorityMap = {
-    prod: 1000,
-    test: 100,
+    prod: 100,
```
Comment (yqwang-ms) [inline src/rest-server/src/middlewares/v2/hived.js]: Comment our allowed range here:
[-1, 126]

Comment (abuccts) [reply src/rest-server/src/middlewares/v2/hived.js]: updated

File: src/rest-server/src/models/v2/job/k8s.js [hunk 2]
```diff
@@ -282,6 +282,15 @@ const generateTaskRole = (taskRole, labels, config) => {
     gangAllocation = 'false';
     retryPolicy.fancyRetryPolicy = true;
   }
+  // calculate pod priority
+  // reference: https://github.com/microsoft/pai/issues/3704
+  let jobPriority = 127;
+  if (launcherConfig.enabledHived) {
+    jobPriority = 126 - parseInt(config.taskRoles[taskRole].hivedPodSpec.priority);
```
Comment (yqwang-ms) [inline src/rest-server/src/models/v2/job/k8s.js]: >priority [](start = 73, length = 8)

What if user does not specify hivedPodSpec.priority?

File: src/rest-server/src/models/v2/job/k8s.js [hunk 3]
```diff
@@ -282,6 +282,15 @@ const generateTaskRole = (taskRole, labels, config) => {
     gangAllocation = 'false';
     retryPolicy.fancyRetryPolicy = true;
   }
+  // calculate pod priority
+  // reference: https://github.com/microsoft/pai/issues/3704
+  let jobPriority = 127;
```
Comment (yqwang-ms) [inline src/rest-server/src/models/v2/job/k8s.js]: >jobPriority [](start = 6, length = 11)

Can we enforce hivedPodSpec.priority in validation [-1, 126]

File: src/rest-server/src/models/v2/job/k8s.js (hunk 2 above)
Comment (abuccts) [reply src/rest-server/src/models/v2/job/k8s.js]: default `hivedPodSpec.priority` is `null`, so default `jobPriority = 126 - parseInt(null)` equals to `126`.

File: src/rest-server/src/models/v2/job/k8s.js (hunk 3 above)
Comment (abuccts) [reply src/rest-server/src/models/v2/job/k8s.js]: it shouldn't be validated here, can be done in hived spec validation.
but at present the priority class uses hard coded values and there's no such validation for the whole hived spec yet.

File: src/rest-server/src/models/v2/job/k8s.js (hunk 2 above)
Comment (fanyangCS) [reply src/rest-server/src/models/v2/job/k8s.js]: isn't 126 too high?

Comment (yqwang-ms) [reply src/rest-server/src/models/v2/job/k8s.js]: Can we explictly specify to default as 0? Max=126 Min=-1 And doc it.

File: src/rest-server/src/models/v2/job/k8s.js (hunk 3 above)
Comment (yqwang-ms) [reply src/rest-server/src/models/v2/job/k8s.js]: Ok, so for safety can we round overflowed jobPriority to be 126, and round undherflowed jobPriority to be -1?

Comment (yqwang-ms) [inline src/rest-server/src/models/v2/job/k8s.js]: >127 [](start = 20, length = 3)

If no hived enabled, let's treat the same as user not specify, i.e. default to jobPriority = 126 - 0

File: src/rest-server/src/models/v2/job/k8s.js [hunk 4]
```diff
@@ -282,6 +282,15 @@ const generateTaskRole = (taskRole, labels, config) => {
...
+  if (launcherConfig.enabledHived) {
+    jobPriority = 126 - parseInt(config.taskRoles[taskRole].hivedPodSpec.priority);
+  }
+  const jobCreationTime = Math.floor(new Date() / 1000) & (Math.pow(2, 23) - 1);
```
Comment (yqwang-ms) [inline src/rest-server/src/models/v2/job/k8s.js]: >new Date() [](start = 37, length = 10)

Is new Date() returns msec instead of sec?

Comment (abuccts) [reply src/rest-server/src/models/v2/job/k8s.js]: yes

File: src/rest-server/src/models/v2/job/k8s.js (hunk 3 above)
Comment (abuccts) [reply src/rest-server/src/models/v2/job/k8s.js]: updated

File: src/rest-server/src/models/v2/job/k8s.js (hunk 2 above)
Comment (abuccts) [reply src/rest-server/src/models/v2/job/k8s.js]: updated comment in hived priority class

File: src/rest-server/src/models/v2/job/k8s.js (hunk 4 above)
Comment (abuccts) [reply src/rest-server/src/models/v2/job/k8s.js]: yes

File: src/rest-server/src/models/v2/job/k8s.js (hunk 3 above)
Comment (yqwang-ms) [reply src/rest-server/src/models/v2/job/k8s.js]: I mean orignal jobPriority to within  [-1, 126], i.e. round hivedPodSpec.priority within it :)

Comment (yqwang-ms) [reply src/rest-server/src/models/v2/job/k8s.js]: I suggest that you can first preprocess the hivedPodSpec.priority
1. If hived not enable, jobPriority = 0
2. else jobPriority = round (hivedPodSpec.priority)
Then reversejobPriority = 126 - jobPriority
Is it more clear?

**Pull request merged**